packages/core/src/Mailbox.ts

```diff
@@ -0,0 +1,36 @@
+export class Mailbox<T> {
+  private events: T[] = [];
+  private index: number = 0;
+
+  public status: 'deferred' | 'idle' | 'processing' = 'deferred';
+
+  public get size(): number {
+    return this.events.length - this.index;
+  }
+
+  public clear(): void {
+    this.events.length = 0;
+    this.index = 0;
+  }
+
+  public enqueue(event: T): void {
+    this.events.push(event);
+  }
+
+  public dequeue(): T | undefined {
+    const event = this.events[this.index];
+
+    if (!event) {
+      return undefined;
+    }
+
+    this.index++;
+
+    if (this.index > this.events.length - 1) {
+      this.events.length = 0;
+      this.index = 0;
+    }
```

Comment:
This strategy doesn't scale because we retain all the queued items up until we drain the queue. The items that were already processed should be immediately disposed because we don't have any use for them from that point in time.

I've reimplemented this mailbox using a simple linked list to avoid this problem altogether:
https://github.com/statelyai/xstate/pull/2916/commits/37547c356cf5f6fe5e073c9c5e1df57e4ddcb43b